osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/security/ReplaceKeyCommandExecutor.java

```diff
@@ -101,58 +107,65 @@ public ActionResponseDto executeBundleAction(final DlmsConnectionManager conn, f
     }
 
-    private SetKeysRequestDto reEncryptKeys(final SetKeysRequestDto setKeysRequestDto) throws FunctionalException {
+    private SetKeysRequestDto decryptRsaKeys(final SetKeysRequestDto setKeysRequestDto) throws FunctionalException {
         final byte[] reEncryptedAuthenticationKey = this.securityKeyService
-                .reEncryptKey(setKeysRequestDto.getAuthenticationKey(), SecurityKeyType.E_METER_AUTHENTICATION);
+                .rsaDecrypt(setKeysRequestDto.getAuthenticationKey());
         final byte[] reEncryptedEncryptionKey = this.securityKeyService
-                .reEncryptKey(setKeysRequestDto.getEncryptionKey(), SecurityKeyType.E_METER_ENCRYPTION);
+                .rsaDecrypt(setKeysRequestDto.getEncryptionKey());
 
         return new SetKeysRequestDto(reEncryptedAuthenticationKey, reEncryptedEncryptionKey);
     }
 
     @Override
     public DlmsDevice execute(final DlmsConnectionManager conn, final DlmsDevice device,
-            final ReplaceKeyCommandExecutor.KeyWrapper keyWrapper) throws OsgpException {
+            final ReplaceKeyCommandExecutor.ReplaceKeyInput keyWrapper) throws OsgpException {
+
+        if (!keyWrapper.isGenerated()) {
+            this.securityKeyService.storeNewKey(device.getDeviceIdentification(), keyWrapper.getSecurityKeyType(),
+                    keyWrapper.getBytes());
+        }
 
-        final DlmsDevice devicePostSave = this.securityKeyService
-                .storeNewKey(device, keyWrapper.getBytes(), keyWrapper.getSecurityKeyType());
-        this.sendToDevice(conn, devicePostSave, keyWrapper);
-        return this.securityKeyService.validateNewKey(devicePostSave, keyWrapper.getSecurityKeyType());
+        this.sendToDevice(conn, device.getDeviceIdentification(), keyWrapper);
+        this.securityKeyService.activateNewKey(device.getDeviceIdentification(), keyWrapper.getSecurityKeyType());
+        return device;
     }
 
     /**
      * Send the key to the device.
      *
      * @param conn
      *         jDLMS connection.
-     * @param device
-     *         Device instance
+     * @param deviceIdentification
+     *         Device identification
      * @param keyWrapper
      *         Key data
      */
-    private void sendToDevice(final DlmsConnectionManager conn, final DlmsDevice device,
-            final ReplaceKeyCommandExecutor.KeyWrapper keyWrapper) throws ProtocolAdapterException {
+    private void sendToDevice(final DlmsConnectionManager conn, String deviceIdentification,
+            final ReplaceKeyCommandExecutor.ReplaceKeyInput keyWrapper) throws ProtocolAdapterException {
 
         try {
-            final byte[] decryptedKey = this.securityKeyService
-                    .decryptKey(keyWrapper.getBytes(), keyWrapper.securityKeyType);
+            final byte[] decryptedKey = keyWrapper.getBytes(); //this.securityKeyService.aesDecryptKey(keyWrapper.getBytes());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This block of commented-out lines of code should be removed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)